draft-ietf-tls-esni.md

```diff
@@ -916,6 +916,30 @@ and session IDs presented by the server. These connections are only used to
 trigger retries, as described in {{rejected-ech}}. This may be implemented, for
 instance, by reporting a failed connection with a dedicated error code.
 
+
+### Impact of Retry on Future Connections
+
+Clients MAY use information learned from a rejected ECH for future
+connections to avoid repeatedly connecting to the same server and
+being forced to retry. However, they MUST handle ECH rejection for
+those connections as if it were a fresh connection, rather than
+enforcing the single retry limit from {{rejected-ech}}.
+
+Clients MUST NOT use ECHConfigs with IP addresses other than
+associated with the DNS queries for the original
+connection. Otherwise, it might become possible for the client to have
```

Comment:
```suggestion
Clients MUST NOT use ECHConfigs with IP addresses other than those
associated with the DNS queries for the original
connection. Otherwise, it might become possible for the client to have
```

However, do you think that you should instead insist on some sort of equality check on something else?  Like the SVCB (or HTTPS) record being for the same QNAME or simply validating that the SVCB record has the same public name?

That latter check seems key, so I might instead say:

```suggestion
Clients MUST NOT reuse ECHConfigs that identify a different public name from
the one used to acquire the IP address of the server. ECHConfigs MUST be freshly
retrieved or taken from a cached SVCB (or HTTPS) record that is within its TTL.
Otherwise, it might become possible for the client to have
```

Reply:
@martinthomson  ping

Reply:
There is no IP in the ECHConfig though, so I'm struggling to understand this.

The trick that I was suggesting was that a server that uses multiple providers could use public name rather than IP to separate those providers.  This is on the presumption that |the set of IP addresses for each provider| > 1 but the number of public names is under their direct control.

Reply:
I have to admit that I have kind of lost track of how the DNS pieces fit together, but...

As I remember, the basic design requirement for multi-CDN was that we had to ensure that there was no way that the client could get ECHConfig for CDN A and the IP address for CDN B, because that just won't work. When I had last checked, that was intended to be accomplished by having the IP address delivered alongside the ECHConfig, which I thought was done with ipv6hint. But maybe I've misremembered or things evolved.

@tfpauly @bemasc?

Reply:
The record can contain v4 and v6 hints, and also is associated with A and AAAA records for that specific service (which would be distinct from the IP addresses for other cdns, etc), that are queried after receiving the HTTPS record.

Reply:
Can you elaborate on how that second process works? By dereferencing the TargetName?

Reply:
Yes, from the target name. See both of these:

https://www.rfc-editor.org/rfc/rfc9460.html#section-7.3

https://www.rfc-editor.org/rfc/rfc9460.html#section-3

Reply:
In SVCB, ECH is a property of the "endpoint", which is represented by one ServiceMode record (out of an RRset that might contain several).  An endpoint also specifies a pool of IP addresses (by the A+AAAA records on the TargetName, or by IP hints in the SvcParams).  As far as SVCB is concerned, ECH can keep state within the scope of a single endpoint, but that state must not leak across to other endpoints of the same "service" (i.e. origin).

This draft doesn't have the concept of an "endpoint", but I think we can define a close enough notion here (without requiring DNS, which is not required for ECH!): call it an "ECHConfig source".

```suggestion
Any persisted information MUST be associated with the ECHConfig source used
to bootstrap the connection, such as DNS SVCB ServiceMode record {{ECH-IN-DNS}}.
Clients MUST limit any sharing of persisted ECH-related state to connections that use
the same ECHConfig source. Otherwise, it might become possible for the client to have
```

Note that "matching ECHConfig" is arguably too broad a scope: if I have two ServiceMode records pointing at separate server pools that initially have the same ECHConfig, I should be able to update one pool without changing the connection behaviors observed by the other one.